### PR TITLE
Allow network-instance to be created only for internal switch communication

### PR DESCRIPTION
**Pull request by naiming-zededa**

### CODE REVIEW

File: pkg/pillar/cmd/zedrouter/dnsmasq.go [hunk 1]
```diff
@@ -128,7 +129,11 @@ func createDnsmasqConfiglet(
 	advertizeRouter := true
 	var router string
 
-	if Ipv4Eid {
+	ifindex, _ := devicenetwork.IfnameToIndex(netconf.Port)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/dnsmasq.go]: Are you also getting the UI/zedcloud to accept 0.0.0.0 as the gateway/router field?
We should fix this as well.

Here I think it makes more sense to check for netconf.Port == "" (which does assume that the UI/zedcloud doesn't feed the device odd things like "0" or "None" when there is no port). But checking for ifindex can cause odd behaviors if the device is in the processing coming back from pciback so I don't think we should depend on that.

File: pkg/pillar/cmd/zedrouter/networkinstance.go [hunk 2]
```diff
@@ -85,6 +85,11 @@ func checkPortAvailable(
 		// XXX Fallback until we have complete Name support in UI
 		portStatus = ctx.deviceNetworkStatus.GetPortByIfName(status.Port)
 		if portStatus == nil {
+			// internal airgap switch case
+			if status.Port == "0" {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/networkinstance.go]: No, that bug in the UI needs to be fixed.
EVE code shouldn't have random workarounds for broken controller/UI code.

File: pkg/pillar/cmd/zedrouter/pbr.go [hunk 3]
```diff
@@ -40,6 +40,11 @@ func PbrRouteAddAll(bridgeName string, port string) error {
 
 	ifindex, err := devicenetwork.IfnameToIndex(port)
 	if err != nil {
+		// for airgap internal switch case, not an error
+		if port == "0" {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/pbr.go]: Ditto

File: pkg/pillar/cmd/zedrouter/dnsmasq.go (hunk 1 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedrouter/dnsmasq.go]: it does create error if putting in 0.0.0.0; but if we allow that in UI, i'm still wondering if define the gateway address may be actually useful from dnsmasq point of view, which our attachment to that bridge in dom0 is the 'gateway' address, but it only offers ip address allocation.

Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/dnsmasq.go]: For the use cases we've seen so far there is a case where a no gateway is useful, which is a Linux domU with two ethernet interfaces where you want the default route only out one of them. For instance, the shopfloor on eth1 is 10.10.10.0/24 and you only want a route for that subnet out eth1.

In that case specifying 0.0.0.0 to mean no default router on that network instance is useful.

**Comment by naiming-zededa** [comment]
have tested this patch with the cloud side changes for Port == "", and also allow user App statically assign the App IP address on the bridge. Additional UI for a radial button of 'not a router' and related API change to handle factory floor network intf can go separately.

File: pkg/pillar/cmd/zedrouter/pbr.go [hunk 4]
```diff
@@ -98,6 +104,12 @@ func PbrRouteAddAll(bridgeName string, port string) error {
 func PbrRouteDeleteAll(bridgeName string, port string) error {
 	log.Infof("PbrRouteDeleteAll(%s, %s)\n", bridgeName, port)
 
+	// for airgap internal switch case
+	if port == "" {
+		log.Infof("PbrRouteAddAll: for internal switch, skip for ACL and Route deletion\n")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/pbr.go]: PbrRouteDeleteAll

Comment (naiming-zededa) [reply pkg/pillar/cmd/zedrouter/pbr.go]: Ok.

**Comment by naiming-zededa** [comment]
@eriknordmark in terms of add the routes in PbrRouteAddAll, at least from the current implementation, we first get all the routes using the ifindex of the external port, then add those routes onto the bnX side. Since this ifindex is not valid in this case, thus we skip for those.

**Comment by eriknordmark** [comment]
> @eriknordmark in terms of add the routes in PbrRouteAddAll, at least from the current implementation, we first get all the routes using the ifindex of the external port, then add those routes onto the bnX side. Since this ifindex is not valid in this case, thus we skip for those.

I thought the code in question takes the routes from bnX (whether it has an external port or not) and adds those to a separate table numbered 500+ifindex(bnX). So it has nothing to do with the external port.

**Comment by naiming-zededa** [comment]
this is the code I'm referring to:
func PbrRouteAddAll(bridgeName string, port string) error {
	log.Infof("PbrRouteAddAll(%s, %s)\n", bridgeName, port)

	// for airgap internal switch case
	if port == "" {
		log.Infof("PbrRouteAddAll: for internal switch, skip for ACL and Route installation\n")
		return nil
	}

	ifindex, err := devicenetwork.IfnameToIndex(port)
	if err != nil {
		errStr := fmt.Sprintf("IfnameToIndex(%s) failed: %s",
			port, err)
		log.Errorln(errStr)
		return errors.New(errStr)
	}
	routes := getAllIPv4Routes(ifindex)

And those 'routes' will be added later into the 500+ifindex table. Unless my understanding of this code is not correct...

**Pull request merged**